homeassistant/util/logging.py

```diff
@@ -164,3 +164,46 @@ def wrapper(*args: Any) -> None:
                 log_exception(*args)
         wrapper_func = wrapper
     return wrapper_func
+
+
+def catch_log_coro_exception(
+        func: Coroutine[Any, Any, Any],
+        format_err: Callable[..., Any],
+        *args: Any) -> Coroutine[Any, Any, Any]:
+    """Decorate a coroutine to catch and log exceptions."""
+    def log_exception(*args: Any) -> None:
+        module_name = inspect.getmodule(inspect.trace()[1][0]).__name__
+        # Do not print the wrapper in the traceback
+        frames = len(inspect.trace()) - 1
+        exc_msg = traceback.format_exc(-frames)
+        friendly_msg = format_err(*args)
+        logging.getLogger(module_name).error('%s\n%s', friendly_msg, exc_msg)
+
+    @wraps(func)  # type: ignore
+    async def coro_wrapper(*args: Any) -> Any:
+        """Catch and log exception."""
+        try:
+            return await func
+        except Exception:  # pylint: disable=broad-except
+            log_exception(*args)
+            return None
+    return coro_wrapper()  # type: ignore
+
+
+def async_create_catching_coro(
+        target: Coroutine) -> Coroutine:
+    """Wrap a coroutine to catch and log exceptions.
+
+    The exception will be logged together with a stacktrace of where the
+    coroutine was wrapped.
+
+    target: target coroutine.
+    """
+    trace = traceback.extract_stack()
+    wrapped_target = catch_log_coro_exception(
+        target, lambda *args:
+        "Exception in {} called from\n {}".format(
+            target.__name__,  # type: ignore
```

Comment:
A coroutine has no __name__ according to MyPy:
`error: "Coroutine[Any, Any, Any]" has no attribute "__name__"`

Reply:
it probably doesn't. It's just an opaque object that can be scheduled.

Reply:
Yeah, but the print works: 
```python
lambda *args:
        "Exception in {} called from\n {}".format(
            target.__name__,
```

Reply:
It seems to have a __name__ attribute, the print is correctly formatted as:
`ERROR:tests.util.test_logging:Exception in job called from`

Where `job` is name of the throwing coroutine function in this test case:
https://github.com/home-assistant/home-assistant/blob/eb21ed2986bb2326ce53875d077ab70aeaacb632/tests/util/test_logging.py#L70-L74

Reply:
oh nice!